### PR TITLE
feat: add Sleep Timer

### PR DESCRIPTION
**Pull request by jhmk**
## Sleep Timer

Adds a Sleep Timer feature accessible from the **Playback > Sleep Timer** submenu.

### Three modes

- **Timed**: Pause/stop after 5, 10, 15, 30, 45, 60, 90, or 120 minutes. Includes a 10-second linear volume fade-out before firing.
- **End of Current Track**: Pause/stop when the currently playing track finishes naturally. Does **not** fire on manual skip/previous — only on natural completion.
- **End of Queue**: Pause/stop when the last track in the playlist finishes.

### Implementation details

- **`SleepTimerManager.swift`** (245 lines): Singleton manager handling all three modes, fade logic, and state management.
- **New `audioTrackDidFinishNaturally` notification**: Posted from `trackDidFinish()` in `AudioEngine.swift` *before* `currentTrack` advances to the next track. This allows `endOfTrack` mode to distinguish natural completion from user skip — addressing a design concern raised in the previous PR review.
- **Menu integration**: Sleep Timer submenu added to the Playback menu in `ContextMenuBuilder.swift` with status display, cancel option, timed presets, and track-aware options. Toggle behavior: selecting the active preset cancels it.
- **Volume safety**: Fade-out detects external volume changes (user adjusts volume during fade) and aborts. Volume is restored to the original level after the timer fires.
- **Thread safety**: All public methods enforce `dispatchPrecondition(condition: .onQueue(.main))`.
- **Session-local state**: Timer state is not persisted across launches (intentional).

### Files changed (3)

| File | Change |
|------|--------|
| `Sources/NullPlayer/App/SleepTimerManager.swift` | **New** — Sleep timer manager |
| `Sources/NullPlayer/App/ContextMenuBuilder.swift` | Added submenu builder + menu actions |
| `Sources/NullPlayer/Audio/AudioEngine.swift` | Added `audioTrackDidFinishNaturally` notification |

### Stats

3 files changed, 362 insertions(+), 2 deletions(-)

## Release Notes

* **New Features**
  * Added Sleep Timer to the Playback menu with three options: timed duration (various minute selections), end of current track, or end of queue
  * Displays current sleep timer status in the menu
  * Toggle an active timer off by selecting the same option again, or cancel anytime from the menu

### CODE REVIEW

**Review by andrewdonheiser — CHANGES_REQUESTED** [review]
Nice work extracting this into a focused PR — the scope is exactly right (3 files, single feature), and the `audioTrackDidFinishNaturally` notification is a clean fix for the critical bug from PR #161. The fade logic, toggle behavior, and thread safety are all solid.

However, two of the three modes have issues that need to be addressed before this can merge. See inline comments for details.

---

## Critical (1)

### endOfQueue mode will never fire

The `trackDidChange` handler checks `currentTrack == nil` to detect queue exhaustion, but `stop()` at end of queue does **not** set `currentTrack = nil` and does **not** post `audioTrackDidChange`. The handler is never invoked. The user sees "Sleep Timer: End of queue" in the menu, but nothing ever happens.

**Fix options:**
- Observe `audioTrackDidFinishNaturally` for endOfQueue too, and check if the finished track is the last in the playlist (`engine.currentIndex >= engine.playlist.count - 1 && !engine.repeatEnabled`)
- Post a new `audioQueueDidExhaust` notification from `trackDidFinish()` when it calls `stop()` at the end of the queue
- Have `stop()` set `currentTrack = nil` when called from a natural queue-end context

## Important (2)

### endOfTrack silently fails when Sweet Fades is enabled

`trackDidFinish()` has a `guard !isCrossfading` early return. Crossfade transitions go through `completeCrossfade()` / `completeStreamingCrossfade()` instead — neither posts `audioTrackDidFinishNaturally`. Users with both Sweet Fades and a sleep timer get no indication it won't work.

**Fix options:**
- Post `audioTrackDidFinishNaturally` in `completeCrossfade()` and `completeStreamingCrossfade()` before advancing `currentTrack`
- Or disable the endOfTrack menu option when Sweet Fades is enabled, with a tooltip explaining why

### No unit tests

`SleepTimerManager` has significant testable logic (fade math, mode handlers, toggle, volume restore). Tests would have caught the endOfQueue bug. Consider adding tests for at minimum: timed fire, endOfTrack on natural notification, endOfQueue on queue exhaust, cancel restores volume during fade, user volume change aborts fade.

## Suggestions

- `tick()` skips `notifyChange()` on early return when fade is cancelled by user volume change — menu countdown misses one update
- `audioTrackDidFinishNaturally` doc comment should state the main-thread delivery guarantee since `SleepTimerManager` depends on it via `dispatchPrecondition`

## Strengths

- Clean fix for PR #161's critical endOfTrack bug — `audioTrackDidFinishNaturally` fires before `currentTrack` advances, no `armedTrackID` workaround needed
- Thread safety verified — both call paths to `trackDidFinish()` dispatch to main
- Fade-out with user-intervention detection is robust
- Toggle behavior (re-select to cancel) works correctly
- Focused 3-file scope, exactly what was asked for

File: Sources/NullPlayer/App/SleepTimerManager.swift [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        timer?.invalidate(); timer = nil
+        tickTimer?.invalidate(); tickTimer = nil
+        self.state = nil
+        fadeStartVolume = nil
+        lastWrittenVolume = nil
+        isFading = false
+        notifyChange()
+    }
+
+    private func applyAction(_ action: Action) {
```
Comment (andrewdonheiser) [inline Sources/NullPlayer/App/SleepTimerManager.swift:200]: **[Critical]** This handler will never be invoked for end-of-queue scenarios.

When the last track finishes, `trackDidFinish()` calls `stop()`, which does NOT set `currentTrack = nil` and does NOT post `audioTrackDidChange`. So this handler is never called, and the sleep timer state is never cleaned up.

You could instead observe `audioTrackDidFinishNaturally` here too, and check whether the finished track is the last in the playlist:

```swift
@objc private func trackDidFinishNaturally(_ note: Notification) {
    dispatchPrecondition(condition: .onQueue(.main))
    guard let state else { return }
    
    if state.mode == .endOfTrack {
        fire()
    } else if state.mode == .endOfQueue {
        let engine = WindowManager.shared.audioEngine
        if engine.currentIndex >= engine.playlist.count - 1 && !engine.repeatEnabled {
            fire()
        }
    }
}
```

Then you can drop the `trackDidChange` observer entirely.

File: Sources/NullPlayer/Audio/AudioEngine.swift [hunk 2]
```diff
@@ -35,7 +35,13 @@ extension Notification.Name {
...
     static let bpmUpdated = Notification.Name("bpmUpdated")
-    
+
+    /// Posted when a track finishes playing naturally (not skipped by user).
+    /// Fired from `trackDidFinish()` BEFORE `currentTrack` advances to the next track.
+    /// userInfo contains: "track" (Track) - the track that just finished
+    /// Used by SleepTimerManager to distinguish natural completion from manual skip.
+    static let audioTrackDidFinishNaturally = Notification.Name("audioTrackDidFinishNaturally")
+
 }
```
Comment (andrewdonheiser) [inline Sources/NullPlayer/Audio/AudioEngine.swift:46]: **[Important]** This notification is never posted when crossfade (Sweet Fades) is active.

`trackDidFinish()` has a `guard !isCrossfading` early return above — crossfade transitions go through `completeCrossfade()` / `completeStreamingCrossfade()` instead, and neither posts this notification. 

If a user has both Sweet Fades and "End of Current Track" sleep timer active, the timer silently does nothing. Consider also posting `audioTrackDidFinishNaturally` from those crossfade completion methods before `currentTrack` advances.

**Comment by ad-repo** [comment]
Addressed all review issues. The fixes are on `feature/sleep-timer` at commit `32b9771`. Summary:

**Critical**
- **endOfQueue mode never fired**: Replaced the broken `audioTrackDidChange` observation (which never fires when `stop()` is called at queue end) with a new `audioQueueDidExhaust` notification. It's posted from a new `stopAfterQueueExhausted()` helper that replaces all three `stop()` calls at natural queue-end sites in `trackDidFinish()` — sequential end-of-list, shuffle exhausted, and repeat+shuffle exhausted.

**Important**
- **endOfTrack fails with Sweet Fades**: Added `audioTrackDidFinishNaturally` post in both `completeCrossfade()` and `completeStreamingCrossfade()` before advancing `currentTrack`, so the sleep timer fires correctly during crossfade transitions.

**CodeRabbit**
- **Timer run loop mode**: Changed both `Timer.scheduledTimer` calls to `Timer(timeInterval:...)` + `RunLoop.main.add(..., forMode: .common)` so fire/tick callbacks continue during menu tracking.
- **Background thread crash**: Changed both notification handlers from `dispatchPrecondition` (crash) to `DispatchQueue.main.async` (safe marshal).
- **Menu items enabled without active track**: `endOfTrack` and `endOfQueue` items now check `audioEngine.currentTrack != nil` and set `isEnabled` accordingly.

**Suggestion**
- `tick()` now calls `notifyChange()` before the early return when a user volume change aborts the fade, so the menu countdown updates immediately.

**Comment by ad-repo** [comment]
Closing in favour of #180, which incorporates your implementation plus all fixes from the review (endOfQueue never firing, endOfTrack failing with Sweet Fades, timer run loop mode, background thread crash, menu items enabled without a track). Also adds 5/8/12 hour options, a live countdown in the submenu, and a Zz indicator in the modern window info row. Thanks for the solid base — the core design and fade logic carried over intact.
